[SR] Support for SLM on Cloud

alisonelizabeth opened the pull request:
This PR adds support for SLM on Cloud.

Fixes https://github.com/elastic/kibana/issues/48402

### How to test
You will need to mock the cloud environment to test this PR.

1. Specify the `path.repo` variable when starting ES:

   ``` 
   yarn es snapshot --license trial -E path.repo=/tmp/es-backups
   ```

2. Add the `cluster.metadata.managed_repository` and `cluster.metadata.managed_policies` settings via Console:

   ```
   PUT /_cluster/settings
   {
     "persistent": {
       "cluster.metadata.managed_repository": "found-snapshots",
       "cluster.metadata.managed_policies": ["managed-policy"]
     }
   }
   ```
3. Create a repository with the same name as your setting value (`found-snapshots`)
4. Create a policy with the same name as your setting value (`managed-policy`)

### What to test
- [Policies tab] A user should not be permitted to delete a managed SLM policy via the table or the details flyout.
- [Policies tab] When creating a policy, a user should not be allowed to select a repo that is already associated with another policy
   - You'll need to set up one policy using the managed repo specified in step 2, and then attempt to create a second policy. Note: validation occurs on blur.
- [Policies tab] If a user edits a managed policy, the `indices` toggle on the second step ("Settings") should be disabled.
- [Policies tab] A "managed" badge should appear in the table next to any managed policies.
- [Policies tab] A warning callout should appear on the details panel and edit flow for managed policies.
- [Snapshots tab] A user should not be able to delete the last successful snapshot associated with a managed repository via the table or the details flyout.
- Confirm there are no issues when using SR without "managed" policies or repos.

<details>
<summary>Screenshots</summary>
<img width="577" alt="Screen Shot 2019-11-22 at 2 20 21 PM" src="https://user-images.githubusercontent.com/5226211/69455019-7c6a7200-0d35-11ea-9243-07eddf5f36bd.png">
<img width="1237" alt="Screen Shot 2019-11-22 at 2 22 52 PM" src="https://user-images.githubusercontent.com/5226211/69455020-7c6a7200-0d35-11ea-8583-416f2c30eb0c.png">
<img width="1254" alt="Screen Shot 2019-11-22 at 2 23 08 PM" src="https://user-images.githubusercontent.com/5226211/69455021-7c6a7200-0d35-11ea-975f-ab38abefb6b5.png">
<img width="1250" alt="Screen Shot 2019-11-22 at 2 26 19 PM" src="https://user-images.githubusercontent.com/5226211/69455022-7d030880-0d35-11ea-8967-b218f43129fe.png">
<img width="631" alt="Screen Shot 2019-11-22 at 2 27 22 PM" src="https://user-images.githubusercontent.com/5226211/69455023-7d030880-0d35-11ea-8ab2-cf21ea27128d.png">
<img width="1188" alt="Screen Shot 2019-11-22 at 2 29 42 PM" src="https://user-images.githubusercontent.com/5226211/69455024-7d030880-0d35-11ea-920c-e07551778069.png">

</details>

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10592/)
* Commit: 939f88bd9c188865f20084c149389819790b6b75

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11488/)
* Commit: 20a6c04a711a33a587f9db590e4a7cd0b050b587

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11595/)
* Commit: e92dd955ce3a2f873309c16840cb4dcd31c17da2

alisonelizabeth commented:
@elasticmachine merge upstream

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11618/)
* Commit: d41bde45aa1a08ed6408f4d34d759092958d88f9

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11628/)
* Commit: caae91258b51968824d76db259250cb259c5bb36

jloleysens commented on x-pack/legacy/plugins/snapshot_restore/public/app/components/policy_form/steps/step_settings.tsx:
```diff
@@ -78,216 +79,239 @@ export const PolicyStepSettings: React.FunctionComponent<StepProps> = ({
+        fullWidth
+      >
+        <EuiFormRow hasEmptyLabelSpace fullWidth describedByIds={['indicesDescription']}>
+          <Fragment>
```
Doesn't look like `Fragment` is needed here (same for points below)? I see it was pre-existing.

alisonelizabeth replied:
This actually is needed. `<EuiFormRow>` expects a single child.

jloleysens commented on x-pack/legacy/plugins/snapshot_restore/public/app/sections/home/policy_list/policy_details/tabs/tab_summary.tsx:
```diff
@@ -130,6 +132,22 @@ export const TabSummary: React.FunctionComponent<Props> = ({ policy }) => {
   return (
     <Fragment>
+      {isManagedPolicy ? (
+        <>
```
Looks like we are already inside a fragment (L134) so can remove this.

alisonelizabeth replied:
I think this is still needed here, as we're also rendering `EuiSpacer`

jloleysens commented on x-pack/legacy/plugins/snapshot_restore/public/app/sections/home/snapshot_list/snapshot_table/snapshot_table.tsx:
```diff
@@ -33,6 +33,19 @@ interface Props {
   onSnapshotDeleted: (snapshotsDeleted: Array<{ snapshot: string; repository: string }>) => void;
 }
 
+const getLastSuccessfulManagedSnapshot = (snapshots: SnapshotDetails[]) => {
```
```suggestion
const getLastSuccessfulManagedSnapshot = (snapshots: SnapshotDetails[]): SnapshotDetails | undefined => {
```

Looks like below uses of `Boolean(lastSuccessfulManagedSnapshot)` should also become just `lastSuccessfulManagedSnapshot` so that TS can infer conditional type correctly.

jloleysens commented on x-pack/legacy/plugins/snapshot_restore/public/app/sections/policy_edit/policy_edit.tsx:
```diff
@@ -182,17 +183,35 @@ export const PolicyEdit: React.FunctionComponent<RouteComponentProps<MatchParams
-        onSave={onSave}
-        onCancel={onCancel}
-      />
+      <>
```
Can remove this outer fragment.

alisonelizabeth replied:
I think this one is still needed as well

jloleysens commented on x-pack/legacy/plugins/snapshot_restore/public/app/services/validation/validate_policy.ts:
```diff
@@ -15,10 +15,14 @@ const isStringEmpty = (str: string | null): boolean => {
-export const validatePolicy = (policy: SlmPolicyPayload): PolicyValidation => {
+export const validatePolicy = (
+  policy: SlmPolicyPayload,
+  validationHelperData: any
```
Can we add typings here 😊 ?

jloleysens commented on x-pack/legacy/plugins/snapshot_restore/server/routes/api/repositories.ts:
```diff
@@ -34,14 +39,19 @@ export function registerRepositoriesRoutes(router: Router, plugins: Plugins) {
+
 export const getAllHandler: RouterRouteHandler = async (
-  req,
+  _req,
```
Just curious, why was this renamed?

alisonelizabeth replied:
Ah, I think this might have been auto-fixed by my editor since it is unused. I reverted the change.

jloleysens commented on x-pack/legacy/plugins/snapshot_restore/public/app/sections/home/snapshot_list/snapshot_table/snapshot_table.tsx:
```diff
@@ -193,29 +208,32 @@ export const SnapshotTable: React.FunctionComponent<Props> = ({
                         {
                           defaultMessage:
-                            'You cannot delete a snapshot stored in a managed repository.',
+                            'You must store at least the last successful snapshot in a managed repository.',
```
```suggestion
                            'You must store the last successful snapshot in a managed repository.',
```

jloleysens commented:
Feedback after manual testing:

### General UX Suggestions

Given space constraints in the policies table, should we consider condensing information with the `EuiIcon` + `EuiToolTip` pattern instead of `EuiBadge` - will probably need to update the copy:

#### Badge

<img width="244" alt="Screenshot 2019-11-25 at 11 40 14" src="https://user-images.githubusercontent.com/8155004/69533834-df981680-0f78-11ea-8bd7-2ff35bebc60d.png">

#### Icon

<img width="352" alt="Screenshot 2019-11-25 at 11 39 48" src="https://user-images.githubusercontent.com/8155004/69533847-eaeb4200-0f78-11ea-9ab8-49b77fa64186.png">

### Editing managed policies

> If a user edits a managed policy, the indices toggle on the second step ("Settings") should be disabled

I was able to reproduce the expected behaviour and saw the callout (nice!). But on blur I still got this validation: 

<img width="540" alt="Screenshot 2019-11-25 at 11 41 46" src="https://user-images.githubusercontent.com/8155004/69533951-2554df00-0f79-11ea-9618-083233014909.png">

Is this expected? Given that I'm editing a pre-existing policy not creating a new one?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12239/)
* Commit: a874f8bdac8ec7573e4efd41300d3ef4323be06a

alisonelizabeth commented:
@elasticmachine merge upstream

elasticmachine commented:
merge conflict between base and head

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13180/)
* Commit: 2d9b6ebe95529c64cb5ced3bc91bd9472abbe767

alisonelizabeth commented:
@jloleysens thanks for the review!

> Given space constraints in the policies table, should we consider condensing information with the EuiIcon + EuiToolTip pattern instead of EuiBadge - will probably need to update the copy:

Good point. I was wondering this as well. I went ahead and made the suggested change. I also switched to the tooltip instead of badge for the managed repository on the repositories tab, so it's consistent.

> I was able to reproduce the expected behaviour and saw the callout (nice!). But on blur I still got this validation:

Great catch! This should be fixed now.

I believe I addressed all of your other comments. Mind taking another look?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13209/)
* Commit: 236a2bb030c8db408c54298401d8eb869d364d7f

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13275/)
* Commit: 307b6056613e864e0a3dfe314d7be5be15f49e46
